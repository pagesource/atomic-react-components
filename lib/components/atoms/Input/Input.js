// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';
import styles from './Input.style';

type Props = {
  id: string,
  className?: string,
  type?: string,
  name: string,
  value?: string | number,
  placeholder?: string,
  ariaLabel?: string,
  elementLocator?: string,
  ref?: ?HTMLInputElement,
};

const Input = React.forwardRef<Props, HTMLInputElement>(
  (
    { id, className, type, name, value, placeholder, ariaLabel, elementLocator, ...others }: Props,
    ref
  ): Node => {
    const propArray = {
      ...others,
      id,
      ref,
      'aria-label': ariaLabel,
      className,
      name,
      type,
      placeholder,
    };
    const propArrayforRadio = {
      ...others,
      id,
      ref,
      'aria-label': ariaLabel,
      className,
      name,
      type,
      placeholder,
      value,
    };
    let finalProp;
    if (type === 'radio') {
      finalProp = propArrayforRadio;
    } else {
      finalProp = propArray;
    }
    return React.createElement('input', finalProp);
  }
);

// $flow-disable-line  eslint-disable-line
Input.defaultProps = {
  placeholder: '',
  ariaLabel: '',
  className: '',
  type: 'text',
  value: '',
  elementLocator: '',
  ref: null,
};

const StyledInput: ComponentType<Props> = styled(Input)`
  ${styles};
`;

export default StyledInput;
export { Input as InputVanilla };
