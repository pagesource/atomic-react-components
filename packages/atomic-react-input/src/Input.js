import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import styles from './Input.style';

const Input = React.forwardRef(
  (
    { id, className, type, name, value, placeholder, ariaLabel, elementLocator, ...others },
    ref
  ) => {
    const propArray = {
      id,
      ref,
      'aria-label': ariaLabel,
      className,
      name,
      type,
      placeholder,
      'data-locator': elementLocator || `input-${type}-${name}`,
      ...others,
    };
    if (type === 'radio') propArray.value = value;
    return React.createElement('input', propArray);
  }
);

Input.propTypes = {
  id: PropTypes.string.isRequired,
  className: PropTypes.string,
  type: PropTypes.string,
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  placeholder: PropTypes.string,
  ariaLabel: PropTypes.string,
  elementLocator: PropTypes.string,
  ref: PropTypes.element,
};

Input.defaultProps = {
  placeholder: '',
  ariaLabel: '',
  className: '',
  type: 'text',
  value: '',
  elementLocator: '',
  ref: null,
};

const StyledInput = styled(Input)`
  ${styles};
`;

export default StyledInput;
export { Input as InputVanilla };
