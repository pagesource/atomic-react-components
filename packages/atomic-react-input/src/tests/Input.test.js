import React from 'react';
import { shallow, render } from 'enzyme';

import { InputVanilla } from '../Input';

const defaultProps = {
  elementLocator: 'my-input-field',
  name: 'my-input',
  type: 'file',
  onChange: () => {},
};

const getWrapperSetup = (props = {}) => {
  const finalProps = { ...defaultProps, ...props };
  return shallow(<InputVanilla {...finalProps} />);
};

describe('Input Component', () => {
  let input = '';
  const sampleProps = {
    elementLocator: 'my-input-field',
    name: 'my-input',
    value: 'abc',
    onChange: () => {},
  };
  test('should render correctly', () => {
    input = shallow(<InputVanilla {...sampleProps} />);

    expect(input).toMatchSnapshot();
  });
  test('should check the data attribute of input', () => {
    const inputVanilla = render(<InputVanilla {...sampleProps} />);

    expect(inputVanilla.attr('data-locator')).toEqual(sampleProps.elementLocator);
  });
});

describe('File Input Component -> Unit Testing', () => {
  let input = '';

  test('should render file input component correctly', () => {
    input = getWrapperSetup();
    expect(input.find('input').length).toBe(1);
  });

  test('should call function on click of the input field', () => {
    function inputClicked(callback) {
      callback();
    }
    input = getWrapperSetup();
    const changedEvt = jest.fn();
    inputClicked(changedEvt);
    expect(changedEvt).toHaveBeenCalledTimes(1);
  });

  test('should simulate click event of the input field', () => {
    input = getWrapperSetup();
    const wrapper = input.simulate('click');
    expect(wrapper.find('input').length).toBe(1);
  });

  test('should render with prop "multiple" to distinguish multiple or single file', () => {
    const isMultipleProp = {
      multiple: true,
    };
    input = getWrapperSetup(isMultipleProp);
    expect(input.prop('multiple')).toBe(true);
  });

  test('should render input correctly', () => {
    input = getWrapperSetup();
    expect(input.find('input')).toHaveLength(1);
  });
});
