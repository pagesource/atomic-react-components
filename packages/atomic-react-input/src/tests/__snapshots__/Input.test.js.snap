// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should render correctly 1`] = `
<input
  aria-label=""
  className=""
  data-locator="my-input-field"
  name="my-input"
  onChange={[Function]}
  placeholder=""
  type="text"
/>
`;
